src/libsyntax/parse/parser/expr.rs

```diff
@@ -1093,10 +1093,40 @@ impl<'a> Parser<'a> {
         let mut attrs = outer_attrs;
         attrs.extend(self.parse_inner_attributes()?);
 
-        let blk = self.parse_block_tail(lo, blk_mode)?;
+        let blk = if let Some(interp_blk) = self.parse_interp_user_fn_block()? {
+            interp_blk
+        } else {
+            self.parse_block_tail(lo, blk_mode)?
+        };
         Ok(self.mk_expr(blk.span, ExprKind::Block(blk, opt_label), attrs))
     }
 
+    /// In interpreter mode, parses the placeholder for the user fn body as the user-provided
+    /// block (not from the source).
+    /// Outside of interpreter mode, simply returns `Ok(None)`.
+    fn parse_interp_user_fn_block(&mut self) -> PResult<'a, Option<P<Block>>> {
+        if let Some(ref interp_user_fn) = self.interp_user_fn {
+            let super::InterpUserFn {
+                ref placeholder,
+                ref body,
+            } = *interp_user_fn.clone();
+            if let token::Ident(name, _) = self.token.kind {
+                if &*name.as_str() == placeholder {
```

Comment:
This looks like we should be tracking the name as a `Symbol` inside `InterpUserFn`...

Reply:
Why? `Symbol` has associated hygiene.

Reply:
Er, no, Symbols have no hygiene data associated with them, Span does -- an `Ident` is a combination of Span and Symbol, which is why it has hygiene, for example.

Reply:
My bad! You're right of course. I'll change this.